serialize.go

```diff
@@ -124,21 +130,46 @@ func (j *Jar) allPersistentEntries() []entry {
 	return entries
 }
 
-// lockFileName returns the name of the lock file associated with
-// the given path.
-func lockFileName(path string) string {
-	return path + ".lock"
-}
+const maxRetryDuration = 2 * time.Second
+
+var re = regexp.MustCompile(`[^a-z0-9A-Z]*`)
 
-const maxRetryDuration = 1 * time.Second
+func lockNameFromPath(path string) (string, error) {
+	if path == "" {
+		return "", errors.New("path cannot be empty")
+	}
+	pathBytes := []byte(path)
+	hash := sha256.New()
+	hash.Write(pathBytes)
+	sha := base64.URLEncoding.EncodeToString(hash.Sum(nil))
 
-func lockFile(path string) (io.Closer, error) {
+	if len(path) > 10 {
+		sliceIndex := len(path) - 10
+		path = path[sliceIndex:]
+	}
+	// We start with an alphabetical character, and remove non alphanumeric
+	// characters so that we can't have an invalid name for the lock.
+	final := re.ReplaceAllLiteralString(fmt.Sprintf("L%v%v", sha[:29], path), ``)
```

Comment:
It's certainly not ideal, but as only alphanumeric characters are valid for lock names the hash requires some kind of encoding. It will always be deterministic and the chances of collision are miniscule.